File: 20_有效的括号/isValid.py
class Solution:
    def isValid(self, s: str) -> bool:
        val_dict = {')': '(', '}': '{', ']': '['}
        stack = []
        for i in s:
            if i not in val_dict:
                stack.append(i)
            # 这两个判断不能反过来
            elif stack == [] or stack.pop() != val_dict[i]:
                return False
        return not stack


if __name__ == '__main__':
    s = "[]"
    solu = Solution()
    print(solu.isValid(s))

File: 20_有效的括号/isValid.ts
import ArrayStack from "../其他_栈实现/array-stack"

function isRight(target: string): boolean {
  if(target === ')' || target === ']' || target === '}') {
    return true
  }

  return false
}

function getOrther(target: string): string {
  switch(target) {
    case ')':
      return '('
    case ']':
      return '['
    case '}':
      return '{'
  }

  return ''
}

function isValid(target: string): boolean {
  let resStack = new ArrayStack<string>()

  for(let item of target) {
    
    if(isRight(item)) {
      if(resStack.pop() !== getOrther(item)) {
        return false
      }
    }else {
      resStack.push(item)
    }
  }

  return resStack.isEmpty() ? true : false
}

console.log(isValid('()[[{}]]'))
console.log(isValid('[()[[{}]]'))
